### PR TITLE
[DQT] Backend api re-definition

### PR DESCRIPTION
**Pull request by xlecours**
This is a DRAFT

## Context
With the increasing size of LORIS instances database, the DQT is starting to reach its limits. having all the data transferred to the frontend and filtered client side do not scale well. The transfer of the querying and filtering logic to the backend would reduce the amount of data that need to be send to the browser. Doing so, we want to define an interface (api) for the backend so it can but used by external tools such as the LORIS DQT React application.

## Goals

- [ ] Transfer query filtering logic to the backend to reduce the amount of data transferred to the frontend.
- [ ] Keep memory usage to a minimum to allow for concurrent queries on bigger databases. 
- [ ] Avoid running the same query on the same data multiple time. Keep queries results, with versioning of results, in a temporary location (cache) and permanent location(URL). The cache can be used for history of the user's last queries. The permanent location will serve as a way to access a given query result without having to rerun it.

*Nice to have*
- [ ]  Automatically save last queries run by the user so it can be reloaded in the frontend (history).

## Brief summary of changes
- [x] Adds a draft of the api re-definition for the dqt module.

##
To look at the schema, go to https://editor.swagger.io/ , File > Import URL, paste the raw version of schema.yml
(Until https://github.com/aces/Loris/pull/5354 gets merged)

### CODE REVIEW

File: modules/dqt/static/schema.yml [hunk 1]
```diff
@@ -0,0 +1,346 @@
...
+    get:
+      description: List available queries
+      responses:
+        '200':
+          description: All the queries in 3 lists ; History, Saved, Shared
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/Queries'
+    post:
```
Comment (driusan) [inline modules/dqt/static/schema.yml:26]: When I put this into https://editor.swagger.io/ the examples that it shows have a structure such as

```
     "fields": {
        "category": "string",
        "field": "string",
        "visits": [
          "string"
        ]
      },
```

I don't know enough about swagger to know where the bug is, but I don't see how that can handle a query that has more than 1 field.

File: modules/dqt/static/schema.yml [hunk 2]
```diff
@@ -0,0 +1,346 @@
...
+            $ref: '#/components/schemas/Filters'
+
+      example:
+        ex1: 
+          type: 'group'
+          operator: 'AND'
+          items:
+            - type: object
+              operator: 'equals'
+              category: 'demographics'
```
Comment (driusan) [inline modules/dqt/static/schema.yml:245]: I think once #6936 goes in this (and other places referring to "category"/"field name") are also going to have to know what module the data comes from in some way.

File: modules/dqt/static/schema.yml [hunk 3]
```diff
@@ -0,0 +1,346 @@
...
+      type: object
+      required:
+        - category
+        - field
+      properties:
+        category:
+          type: string
+        field:
+          type: string
+        visits:
```
Comment (driusan) [inline modules/dqt/static/schema.yml:205]: is this required? What about for candidate scoped data?

File: modules/dqt/static/schema.yml (hunk 1 above)
Comment (maltheism) [reply modules/dqt/static/schema.yml:26]: Unsure if helpful but therels also allVisits that get attached like example from console.log:
```
{
  "specimens": {
    "specimen_list_details": {
      "Visit 02": "Visit 02"
    },
    "allVisits": {
      "Visit 02": 1,
      "Visit 01": 0,
      "ALS-TAL-V01": 0,
      "ALS-TAL-V02": 0,
      "ALS-TAL-V03": 0,
      "ALS-TAL-V04": 0,
      "ALS-TAL-VET": 0,
      "ALS-TAL-VOP": 0,
      "ALS-TAL-VT": 0,
      "BENEFIT-Bsl": 0,
      "BENEFIT-Scr": 0,
      "BENEFIT-V01": 0,
      "CEF-Base": 0,
      "CEF-Scr": 0,
      "CEF-W16": 0,
      "CEF-W24": 0,
      "EMP-Base": 0,
      "EMP-M03": 0,
      "EMP-M07": 0,
      "EMP-M08": 0,
      "EMP-M10": 0,
      "EMP-M11": 0,
      "EMP-M12": 0,
      "EMP-Scr": 0,
      "ENV-Base": 0,
      "ENV-EOT": 0,
      "ENV-F/up": 0,
      "ENV-M01": 0,
      "ENV-M02": 0,
      "Genetics-V01": 0,
      "GSK-NOG Bsl": 0,
      "GSK-NOG EW": 0,
      "GSK-NOG FUP": 0,
      "GSK-NOG Scr": 0,
      "GSK-NOG W02": 0,
      "GSK-NOG W03": 0,
      "GSK-NOG W04": 0,
      "GSK-NOG W06": 0,
      "GSK-NOG W10": 0,
      "GSK-NOG W12": 0,
      "GSK-NOG W16": 0,
      "GSK-NOG W20": 0,
      "GSK-NOG W24": 0,
      "GSK-NOG W28": 0,
      "GSK-NOG W32": 0,
      "GSK-NOG W36": 0,
      "GSK-NOG W40": 0,
      "GSK-NOG W44": 0,
      "GSK-NOG W48": 0,
      "LIT-Base": 0,
      "MRI01": 0,
      "MRI01Anonymized": 0,
      "NeuropsychologyEvaluation01": 0,
      "TOSI-ALS-V01": 0,
      "TOSI-MS-V01": 0,
      "V01 (Pre-Inf)": 0,
      "V01-Sal": 0,
      "V02 (D9 Post)": 0,
      "V02-Sal": 0,
      "V03 (D14 Post)": 0,
      "V04 (Pre-Inf)": 0,
      "Visit 01 - NDP": 0,
      "Visit 01 - SDP": 0
    }
  }
}
```

File: modules/dqt/php/provisioners/categoriesinstance.class.inc [hunk 4]
```diff
@@ -0,0 +1,15 @@
+<?php
```
Comment (xlecours) [inline modules/dqt/php/provisioners/categoriesinstance.class.inc:1]: I don't know where to put those classes. There will be many. 
Shold I try to use the DataDictionnary categories instead?

File: modules/dqt/php/provisioners/categoriesprovisioner.class.inc [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+
+        parent::__construct(
+            $loris,
+            $query
+        );
+    }
+
+    public function getInstance($row): DataInstance
+    {
+        $name = json_decode($row,true)['key'][0];
```
Comment (xlecours) [inline modules/dqt/php/provisioners/categoriesprovisioner.class.inc:53]: There should be some checks here

File: modules/dqt/php/provisioners/categoriesinstance.class.inc (hunk 4 above)
Comment (driusan) [reply modules/dqt/php/provisioners/categoriesinstance.class.inc:1]: I think it makes sense to use the Data Dictionary categories.. what's the point in creating a new parallel type?

**Comment by maltheism** [comment]
Hi @xlecours, I'm going to be making a PR on Loris about the following but I thought to comment here for you to be aware. The DQT's frontend 'table' rendering code had a few performance problems. On the project CBIGR I made a separate table component and stripped out code that's not being used for the DQT such as for filtering and while removing/changing existing code. Latest PR https://github.com/aces/CBIGR/pull/175/files as reference and previous PRs were merged with fixes.

**Pull request closed without merging**